### PR TITLE
Handle Suspended Resolution

### PR DESCRIPTION
**Pull request by benwilson512**
Hey @sikanhe.

I've been looking at this library more now that GraphiQL Playground supports apollo tracing, thanks so much for taking a stab at getting this going!

I've made a couple changes here so that it can properly support batched data, I'll annotate specific spots with details.

image
Batching at work ^

### CODE REVIEW

File: lib/apollo_tracing/middleware.ex [hunk 1]
```diff
@@ -3,52 +3,38 @@ defmodule ApolloTracing.Middleware do
...
-      apollo_tracing_start_time: start_mono_time,
-      apollo_tracing: %Schema{
-        execution: %Execution{
-          resolvers: resolvers_so_far
-        }
-      }
-    }} = res
-
+  # if there isn't an `ApolloTracing` flag set then we aren't actually doing any tracing
+  def call(%Resolution{acc: %{apollo_tracing_start_time: start_mono_time}, state: :unresolved} = res, _config) do
```
Comment (benwilson512) [inline lib/apollo_tracing/middleware.ex:11]: The pattern match here, and the fall through pattern on line 26 handle situations where the user doesn't actually want to run tracing for a given run.

File: lib/apollo_tracing/middleware.ex [hunk 2]
```diff
@@ -3,52 +3,38 @@ defmodule ApolloTracing.Middleware do
...
 
-    res = put_in(
-      res.acc.apollo_tracing.execution.resolvers,
-      [resolver | resolvers_so_far]
-    )
-
-    %{res | middleware:
-       res.middleware ++ [{{__MODULE__, :after_field}, [start_time: now]}]
+    %{res |
+      extensions: Map.put(res.extensions, __MODULE__, resolver),
```
Comment (benwilson512) [inline lib/apollo_tracing/middleware.ex:22]: In the previous way this worked every `call` would basically handle the last item on the `execution.resolvers` stack. Problems occur with the following type of situation:

```
{
  foo1: getFoo(id: 1) { name }
  foo2: getFoo(id: 2) { name }
}
```

If the `getFoo` field is using batching, then `call/2` in this module will run for `foo1` field, but then `foo1` gets suspended. Absinthe moves on to `foo2` BEFORE `foo1` completes, and thus before the `after_field` call happens on `foo1`. This causes a variety of problems.

In any case, I solved it by storing the `resolver` info under the extensions key of the resolution struct. This extensions key is not automatically aggregated, and thus is a handy spot to store data you want to just keep for this specific field. I store it under the `__MODULE__` key just to avoid name conflicts.

File: lib/apollo_tracing/middleware.ex [hunk 3]
```diff
@@ -3,52 +3,38 @@ defmodule ApolloTracing.Middleware do
...
+  end
 
   # Called after each resolution to calculate the duration
   def after_field(%Resolution{state: :resolved} = res, [start_time: start_time]) do
-    %{acc: %{
-      apollo_tracing: %Schema{
-        execution: %Execution{resolvers: [resolver | prev_resolvers]}
-      }
-    }} = res
+    %{extensions: %{__MODULE__ => resolver}} = res
```
Comment (benwilson512) [inline lib/apollo_tracing/middleware.ex:32]: Now that the field has actually completed, I pull the resolver info for THIS specific field off of its extensions, and then on line 38 I can push that onto your `execution.resolvers` stack.

File: lib/apollo_tracing/pipeline.ex [hunk 4]
```diff
@@ -5,17 +5,19 @@ defmodule ApolloTracing.Pipeline do
...
-      {:module, absinthe_plug} ->
-        absinthe_plug.default_pipeline(config, pipeline_opts)
-       |> add_phases()
-      _ ->
-        raise RuntimeError, """
-          You don't have Plug loaded, please use
-          ApolloTracing.Pipeline.default(absinthe_schema, pipeline_opts)
-          to produce a pipeline without Plug specific phases
-        """
+  if Code.ensure_loaded?(Absinthe.Plug) do
```
Comment (benwilson512) [inline lib/apollo_tracing/pipeline.ex:8]: This is unrelated to the other changes. `Code.ensure_loaded` can be a VERY expensive call, and should basically never happen at runtime. Always do the check at compile time and have it generate different functions.

File: lib/apollo_tracing/middleware.ex (hunk 1 above)
Comment (benwilson512) [reply lib/apollo_tracing/middleware.ex:11]: @athal7 This also handles the issue you were facing with the `prime` function.

**Comment by benwilson512** [comment]
Notably this does NOT fix subscriptions, but that'll be another PR.

**Comment by sikanhe** [comment]
@benwilson512 That's amazing. Thank you and @athal7  so much for helping out with this.

**Comment by benwilson512** [comment]
Thanks! Hold off on doing a release, got another PR or two that may be
incoming.
